i3lock.c

```diff
@@ -645,6 +645,120 @@ void handle_screen_resize(void) {
     redraw_screen();
 }
 
+static int read_raw_image_native(uint32_t *dest, FILE *src, int width, int height, int stride) {
+    int count = 0;
+    for (int y = 0; y < height; y++) {
+        size_t n = fread(&dest[y * stride], 1, width * 4, src);
+        count += n;
+        if (n < (size_t)(width * 4))
+            break;
+    }
+
+    return count;
+}
+
+static int read_raw_image_rgb(uint32_t *dest, FILE *src, int width, int height, int stride) {
+    unsigned char *buf = malloc(width * 3);
+    if (buf == NULL)
+        return -1;
+
+    int count = 0;
+    for (int y = 0; y < height; y++) {
+        size_t n = fread(buf, 1, width * 3, src);
+        count += n;
+        if (n < (size_t)(width * 3))
+            break;
+
+        for (int x = 0; x < width; ++x) {
+            int idx = x * 3;
+            dest[y * stride + x] = 0 |
+                                   (buf[idx + 0] << 16) |
+                                   (buf[idx + 1] << 8) |
+                                   (buf[idx + 2]);
+        }
+    }
+
+    free(buf);
+    return count;
+}
+
+static cairo_surface_t *read_raw_image(const char *image_path, const char *image_raw_format) {
+    cairo_surface_t *img;
+
+    /* Parse format as <width>x<height>[:<format>] */
+    int w = -1, h = -1;
+    char format[] = "native";
+    char *fmtstring = NULL;
+    if (asprintf(&fmtstring, "%%ix%%i:%%%lds", sizeof(format) - 1) == -1) {
```

Comment:
I noticed that there was some discussion regarding this, but... really?

Reply:
We couldn’t find a way to express `strlen("native")` at compile time within a format string. Do you know of one?

Reply:
Not really. I guess I would do it the other way around. I feel like that way might be easier to understand. (50 is just some number larger than `strlen("native")`)
```C
#define MY_LENGTH 50
#define STRINGIFY1(x) #x
#define STRINGIFY(x) STRINGIFY1(x)
	char buffer[MY_LENGTH+1];
	sscanf(string, "%" STRINGIFY(MY_LENGTH) "s", buffer);
```

Reply:
I believe this is essentially the last remaining point before I squash commits again and things are ready to merge (unless there are any issues I haven't noticed with my new changes of course). I don't particularly care if we leave it as it is, revert to how I originally had it, or do something like what psychon suggests (which is fairly similar to how it was originally, except that the length is set by a macro).

Reply:
Let’s go with @psychon’s suggestion, but using a buffer size that is only as large as we need.

Reply:
Alright, I changed it: https://github.com/i3/i3lock/pull/226/commits/80e6215e6f45e4615f11558bcc13e2a04f4f880a